### DISCUSSION TITLE
VBR detection

### DISCUSSION THREAD

**Discussion by naanlizard** [OvenMediaLabs/OvenMediaEngine#1656] [Ideas]
With the new API responses that include bitrate (configured and actual), would it be possible to programmatically detect whether a user is using VBR as their rate control? 

I'd assume something like "if bitrate avg is significantly below configured average, VBR likely"

But I'm not sure how dropped frames would show here

This would be useful for informing users to switch to or away from VBR, we recommend it but others usually don't I think

**Comment by getroot** [comment]
A method to detect VBR (Variable Bitrate) involves observing the characteristic of fluctuating bitrates over time. Unlike CBR (Constant Bitrate), where the bitrate remains constant, VBR shows significant variations in bitrate. Therefore, by monitoring the bitrate and identifying substantial fluctuations, one can detect VBR.

However, this can be influenced by the encoder's implementation, its performance, or factors such as packet loss. Since these factors also affect video quality, it is a good idea to issue a warning like the one below:

"You might be encoding in VBR mode, or the video quality is inconsistent."

**Reply by naanlizard** [reply]
>However, this can be influenced by the encoder's implementation, its performance, or factors such as packet loss. Since these factors also affect video quality, it is a good idea to issue a warning like the one below:

Exactly my concern, I was hoping you'd have some idea how to feasibly do this

Note, in our use case, we WANT users to use VBR. VBR is good for us. 

Does OME have any way to detect dropped frames or packet loss? If we can do that, and see

1. No packet loss or dropped frames
2. Variable bit rate (over some % threshold)

= probably VBR

Separately, if we can detect packet loss or dropped frames, that would be very useful as well!

**Reply by getroot** [reply]
If you are using constant frame rate, then detecting fps continuously will help you detect frame loss. But your encoder is probably rtmp, right? It is tcp based, so there is probably very little frame loss on the network side. If your input is webrtc, then they use variable frame rate.

**Reply by naanlizard** [reply]
Ah for clarity, we have rtmp ingests from arbitrary user-defined encoders - typically x264 CPU in OBS (recommended) or NVENC (also OBS)

Sometimes, users have their bitrate set too high, or have bad internet, etc, which causes dropped frames or packet loss. Generally, FPS can be assumed to be constant though - could we detect encoder issues if the fps is not constant?

**Reply by getroot** [reply]
Yes, if you use OBS your FPS should be constant.

**Reply by getroot** [reply]
even if it is not a frame loss situation, there are cases where the fps statistics gradually decrease because the transmission speed is slow and the frames arrive late. Therefore, I think the idea of ​​detecting abnormal situations by detecting non-constant fps is a good idea.

**Reply by naanlizard** [reply]
Is that something you would suggest users implement, or something that could be returned by OME? e.g. InconsistentFPS: {true|false|unknown} in the API

An aside, where are the new statistics that were added in 0.16.6 found? e.g. bitrateConf - is gitbook not updated? The only place I see "bitrateConf" is in the multiplex feature API docs https://airensoft.gitbook.io/ovenmediaengine/rest-api/v1/virtualhost/application/scheduledchannel-api-1#post-v1-vhosts-vhost-apps-app-multiplexchannels

**Reply by getroot** [reply]
We do not update all detailed API items in Gitbook. Check the results by querying with the GET v1/vhosts/vhost name/apps/app name/streams/stream name API. I think you can implement what we discussed by using its fps-related statistical information.
